Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,617 @@
+    (U_symm : SymmetricRel U) {m : ℕ} (n : ℕ) {s : Finset X} (h : IsDynCoverOf T F U m s) :
+    ∃ t : Finset X, IsDynCoverOf T F (U ○ U) (m * n) t ∧ t.card ≤ s.card ^ n := by
+  classical
+  rcases F.eq_empty_or_nonempty with (rfl | F_nemp)
+  · use ∅; simp
+  have _ : Nonempty X := nonempty_of_exists F_nemp
+  have s_nemp := h.nonempty F_nemp
+  rcases F_nemp with ⟨x, x_F⟩
+  rcases m.eq_zero_or_pos with (rfl | m_pos)
+  · use {x}
+    simp only [zero_mul, Finset.coe_singleton, Finset.card_singleton]
+    exact And.intro (isDynCoverOf_zero T F (U ○ U) (singleton_nonempty x))
+      <| one_le_pow_of_one_le' (Nat.one_le_of_lt (Finset.Nonempty.card_pos s_nemp)) n
+  have : ∀ β : Fin n → s, ∃ y : X, (⋂ k : Fin n, T^[m * k] ⁻¹' ball (β k) (dynEntourage T U m)) ⊆
+      ball y (dynEntourage T (U ○ U) (m * n)) := by
+    intro t
+    rcases (⋂ k : Fin n, T^[m * k] ⁻¹' ball (t k) (dynEntourage T U m)).eq_empty_or_nonempty
+      with (inter_empt | inter_nemp)
+    · rw [inter_empt]
+      use x
+      exact empty_subset _
+    · rcases inter_nemp with ⟨y, y_int⟩
+      use y
+      intro z z_int
+      simp only [ball, dynEntourage, Prod.map_iterate, mem_preimage, mem_iInter,
+        Prod.map_apply] at y_int z_int ⊢
+      intro k k_mn
+      replace k_mn := Nat.div_lt_of_lt_mul k_mn
+      specialize z_int ⟨(k / m), k_mn⟩ (k % m) (Nat.mod_lt k m_pos)
+      specialize y_int ⟨(k / m), k_mn⟩ (k % m) (Nat.mod_lt k m_pos)
+      rw [← Function.iterate_add_apply T (k % m) (m * (k / m)), Nat.mod_add_div k m] at y_int z_int
+      exact mem_comp_of_mem_ball U_symm y_int z_int
+  choose! dyncover h_dyncover using this
+  let sn := range dyncover
+  have _ := fintypeRange dyncover
+  use sn.toFinset
+  constructor
+  · rw [Finset.coe_nonempty] at s_nemp
+    have _ : Nonempty s := Finset.Nonempty.coe_sort s_nemp
+    intro y y_F
+    have key : ∀ k : Fin n, ∃ z : s, y ∈ T^[m * k] ⁻¹' ball z (dynEntourage T U m) := by
+      intro k
+      have := h (MapsTo.iterate F_inv (m * k) y_F)
+      simp only [Finset.coe_sort_coe, mem_iUnion, Subtype.exists, exists_prop] at this
+      rcases this with ⟨z, z_s, hz⟩
+      use ⟨z, z_s⟩
+      exact hz
+    choose! t ht using key
+    specialize h_dyncover t
+    simp only [Finset.coe_sort_coe, mem_iUnion, Subtype.exists, toFinset_range,
+      Finset.mem_image, Finset.mem_univ, true_and, exists_prop, exists_exists_eq_and, sn]
+    simp only [mem_preimage, Subtype.forall, Finset.mem_range] at ht
+    use dyncover t
+    simp only [Finset.coe_image, Finset.coe_univ, image_univ, mem_range, exists_apply_eq_apply,
+      true_and]
+    apply h_dyncover
+    simp only [mem_iInter, mem_preimage, Finset.mem_range]
+    exact ht
+  · rw [toFinset_card]
+    apply le_trans (Fintype.card_range_le dyncover)
+    simp only [Fintype.card_fun, Fintype.card_coe, Fintype.card_fin, le_refl]
+
+lemma isDynCoverOf_compact_continuous [UniformSpace X] {T : X → X} (h : UniformContinuous T)
```

Comment:
If the lemma involves `IsCompact` and `UniformContinuous`, the lemma name should probably be `exists_isDynCoverOf_of_isCompact_uniformContinuous` (and maybe swap the order of the assumptions to match the name of the lemma).

Reply:
Done. Thanks.